engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -102,29 +122,128 @@ public final ImmutableTableKey getKey() {
+                    if (removedLocation != null) {
+                        removedImmutableKeys.add(toKeyImmutable(locationKey));
+                    }
+                }
+            }
+        }
+
+        if (subscriptions != null) {
+            synchronized (subscriptions) {
+                // Push the notifications to the subscribers.
+                if ((!addedImmutableKeys.isEmpty() || !removedImmutableKeys.isEmpty())
+                        && subscriptions.deliverNotification(
+                                Listener::handleTableLocationKeysUpdate,
+                                addedImmutableKeys,
+                                removedImmutableKeys,
+                                true)) {
+                    onEmpty();
+                }
+            }
+        }
     }
 
+    /**
+     * Deliver a possibly-new key.
+     *
+     * @param locationKey The new key
+     * @apiNote This method is intended to be used by subclasses or by tightly-coupled discovery tools.
+     */
     protected final void handleTableLocationKeyAdded(@NotNull final TableLocationKey locationKey) {
         handleTableLocationKeyAdded(locationKey, null);
     }
 
     /**
-     * Deliver a possibly-new key.
+     * Deliver a possibly-new key, optionally as part of a transaction.
      *
      * @param locationKey The new key
-     * @param transactionToken The token identifying the transaction
+     * @param transactionToken The token identifying the transaction (or null if not part of a transaction)
      * @apiNote This method is intended to be used by subclasses or by tightly-coupled discovery tools.
      */
     protected final void handleTableLocationKeyAdded(
             @NotNull final TableLocationKey locationKey,
             @Nullable final Object transactionToken) {
+
         if (!supportsSubscriptions()) {
             tableLocations.putIfAbsent(locationKey, TableLocationKey::makeImmutable);
-            visitLocationKey(toKeyImmutable(locationKey));
+            visitLocationKey(locationKey);
+            return;
+        }
+
+        if (transactionToken != null) {
+            // When adding a location in a transaction, check for logical consistency.
+            // 1. If the location was already added in this transaction, we have a problem. A transaction should not
+            // add the same location twice.
+            // 2. If the location was already removed in this transaction, we have a `replace` operation which is not a
+            // logical error (although it may not be supported by all consumers).
+            synchronized (transactionLock) {
+                final Transaction transaction = transactions.get(transactionToken);
+                if (transaction == null) {
+                    throw new IllegalStateException(
+                            "No transaction with token " + transactionToken + " is currently open.");
+                }
+                transaction.addlocationKey(locationKey);
```

Comment:
This probably need not be under the lock. Or it can be under a lock on the transaction.